File: LiveSendGift/LiveGiftShowView/LiveGiftShowCustom.m
//
//  LiveGiftShowCustom.m
//  LiveSendGift
//
//  Created by Jonhory on 2016/12/4.
//  Copyright © 2016年 com.wujh. All rights reserved.
//

#import "LiveGiftShowCustom.h"
#import "LiveGiftShowView.h"

static CGFloat const kGiftViewMargin = 50.0;/**< 两个弹幕之间的高度差 */
static NSString * const kGiftViewRemoved = @"kGiftViewRemoved";/**< 弹幕已移除的key */

@interface LiveGiftShowCustom ()

@property (nonatomic ,strong) NSLayoutConstraint * heightConstraint;/**< 容器高度约束，随 maxRailwayCount 更新 */
@property (nonatomic ,strong) NSMutableDictionary * showViewDict;/**< key([self getDictKey]):value(LiveGiftShowView*) */
// 用来记录模型的顺序
@property (nonatomic ,strong) NSMutableArray * showViewArr;/**< [LiveGiftShowView, @"kGiftViewRemoved"] */
// 待展示礼物队列
@property (nonatomic, strong) NSMutableArray <LiveGiftShowModel *> * waitQueueArr;

@end

@implementation LiveGiftShowCustom

#pragma mark - 初始化
+ (instancetype)addToView:(UIView *)superView {
    return [self addToView:superView y:100];
}

+ (instancetype)addToView:(UIView *)superView y:(CGFloat)y {
    LiveGiftShowCustom * v = [[LiveGiftShowCustom alloc]init];
    v.userInteractionEnabled = NO;
    [superView addSubview:v];
    // 布局（V2.0 起用系统 NSLayoutAnchor，移除 Masonry 依赖）
    v.translatesAutoresizingMaskIntoConstraints = NO;
    v.heightConstraint = [v.heightAnchor constraintEqualToConstant:(kViewHeight + kGiftViewMargin) * (v.maxRailwayCount - 1) + kViewHeight];
    [NSLayoutConstraint activateConstraints:@[
        [v.widthAnchor constraintEqualToConstant:kViewWidth],
        v.heightConstraint,
        // 这个可以任意修改
        [v.leftAnchor constraintEqualToAnchor:superView.leftAnchor],
        // 这个参数在的设定应该注意最大礼物数量时不要超出屏幕边界
        [v.topAnchor constraintEqualToAnchor:superView.topAnchor constant:y],
    ]];
    v.backgroundColor = [UIColor clearColor];
    return v;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        // 默认配置（V2.0 起为实例属性，多实例互不影响）
        _kExchangeAnimationTime = 0.25;
        _kAppearAnimationTime = 0.5;
        _addMode = LiveGiftAddModeReplace;
        _maxRailwayCount = 3;
        _railwayCanExchange = YES;
        _showMode = LiveGiftShowModeFromTopToBottom;
        _hiddenMode = LiveGiftHiddenModeRight;
        _appearMode = LiveGiftAppearModeLeft;
        _interfaceDebugEnabled = NO;
    }
    return self;
}

- (void)setMaxRailwayCount:(NSInteger)maxRailwayCount {
    _maxRailwayCount = maxRailwayCount;
    self.heightConstraint.constant = (kViewHeight + kGiftViewMargin) * (maxRailwayCount - 1) + kViewHeight;
}

- (void)addLiveGiftShowModel:(LiveGiftShowModel *)showModel{
    [self addLiveGiftShowModel:showModel showNumber:0];
}

- (void)addLiveGiftShowModel:(LiveGiftShowModel *)showModel showNumber:(NSInteger)showNumber{
    if (!showModel || ![showModel isKindOfClass:[LiveGiftShowModel class]]) {
        return;
    }
    // 线程安全：礼物消息常来自 IM/网络回调线程，统一转到主队列
    if (![NSThread isMainThread]) {
        __weak __typeof(self)weakSelf = self;
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf addLiveGiftShowModel:showModel showNumber:showNumber];
        });
        return;
    }

    LiveGiftShowView * oldShowView = [self.showViewDict objectForKey:[self getDictKey:showModel]];
    
    // 判断是否强制修改显示的数字
    BOOL isResetNumber = showNumber > 0 ? YES : NO;
    
    // 如果不存在旧模型
    if (!oldShowView || ![oldShowView isKindOfClass:[LiveGiftShowView class]]) {
        // 如果当前弹幕数量大于最大限制
        if (self.showViewArr.count >= self.maxRailwayCount) {
            // 判断数组是否包含kGiftViewRemoved , 不包含的时候进行排序
            if (![self.showViewArr containsObject:kGiftViewRemoved]) {
                if (self.addMode == LiveGiftAddModeReplace) {
                    // 排序 最小的时间在第一个
                    NSArray * sortArr = [self.showViewArr sortedArrayUsingComparator:^NSComparisonResult(LiveGiftShowView * obj1, LiveGiftShowView * obj2) {
                        return [obj1.createDate compare:obj2.createDate];
                    }];
                    LiveGiftShowView * oldestView = sortArr.firstObject;
                    // 重置模型
                    [self resetView:oldestView nowModel:showModel isChangeNum:isResetNumber number:showNumber];
                } else {
                    if (showNumber > 0) {
                        showModel.currentNumber = showNumber;
                    }
                    [self addToQueue:showModel];
                }
                return;
            }
        }
        
        // 计算视图Y值
        CGFloat   showViewY = 0;
        if (self.showMode == LiveGiftShowModeFromTopToBottom) {
            showViewY = (kViewHeight + kGiftViewMargin) * [self.showViewDict allKeys].count;
        } else if (self.showMode == LiveGiftShowModeFromBottomToTop) {
            showViewY = - ((kViewHeight + kGiftViewMargin) * [self.showViewDict allKeys].count);
        }
        // 获取已移除的key的index
        NSInteger kRemovedViewIndex = [self.showViewArr indexOfObject:kGiftViewRemoved];
        if ([self.showViewArr containsObject:kGiftViewRemoved]) {
            if (self.showMode == LiveGiftShowModeFromTopToBottom) {
                showViewY = kRemovedViewIndex * (kViewHeight+kGiftViewMargin);
            } else if (self.showMode == LiveGiftShowModeFromBottomToTop) {
                showViewY = - (kRemovedViewIndex * (kViewHeight+kGiftViewMargin));
            }
        }
        // 创建新模型
        CGRect frame = CGRectMake(0, showViewY, 0, 0);
        if (self.appearMode == LiveGiftAppearModeLeft) {
            // 用宿主视图宽度计算离屏起点，iPad 分屏等非全屏容器下 mainScreen 宽度会偏大
            CGFloat containerWidth = self.superview ? CGRectGetWidth(self.superview.bounds) : CGRectGetWidth([UIScreen mainScreen].bounds);
            frame = CGRectMake(-containerWidth, showViewY, 0, 0);
        }
        LiveGiftShowView * newShowView = [[LiveGiftShowView alloc]initWithFrame:frame];
        // 赋值（imageLoader 须在 model 之前，setModel 内会触发图片加载）
        newShowView.imageLoader = self.webImageLoader;
        newShowView.model = showModel;
        newShowView.hiddenMode = self.hiddenMode;
        // 改变礼物数量
        if (isResetNumber) {
            [newShowView resetTimeAndNumberFrom:showNumber];
        }else{
            [newShowView addGiftNumberFrom:1];
        }
        [self appearWith:newShowView];
        
        // 超时移除
        __weak __typeof(self)weakSelf = self;
        newShowView.liveGiftShowViewTimeOut = ^(LiveGiftShowView * willReMoveShowView){
            if (weakSelf.delegate && [weakSelf.delegate respondsToSelector:@selector(giftDidRemove:)]) {
                [weakSelf.delegate giftDidRemove:willReMoveShowView.model];
            }
            // 从数组移除（index 由 sortShowArr 手工维护，做边界防护避免状态不同步时越界崩溃）
            if (willReMoveShowView.index >= 0 && willReMoveShowView.index < weakSelf.showViewArr.count) {
                [weakSelf.showViewArr replaceObjectAtIndex:willReMoveShowView.index withObject:kGiftViewRemoved];
            } else if ([weakSelf.showViewArr containsObject:willReMoveShowView]) {
                NSInteger realIndex = [weakSelf.showViewArr indexOfObject:willReMoveShowView];
                [weakSelf.showViewArr replaceObjectAtIndex:realIndex withObject:kGiftViewRemoved];
            }
            // 从字典移除
            NSString * willReMoveShowViewKey = [weakSelf getDictKey:willReMoveShowView.model];
            [weakSelf.showViewDict removeObjectForKey:willReMoveShowViewKey];
            
            if ([weakSelf isDebug]) {
                WLog(@"移除了第%zi个,移除后数组 = %@ ,词典 = %@",willReMoveShowView.index,weakSelf.showViewArr,weakSelf.showViewDict);
            }
            
            if (weakSelf.railwayCanExchange) {
                // 比较数量大小排序
                [weakSelf sortShowArr];
                [weakSelf resetY];
            }
            
            if (weakSelf.addMode == LiveGiftAddModeQueue) {
                [weakSelf showWaitView];
            } else if (weakSelf.addMode == LiveGiftAddModeReplace) {
                if (willReMoveShowView.model.animatedTimer) {
                    dispatch_cancel(willReMoveShowView.model.animatedTimer);
                    willReMoveShowView.model.animatedTimer = nil;
                }
            }
        };
        
        [self addSubview:newShowView];
        
        // 加入数组
        if ([self.showViewArr containsObject:kGiftViewRemoved]) {
            newShowView.index = kRemovedViewIndex;
            [self.showViewArr replaceObjectAtIndex:kRemovedViewIndex withObject:newShowView];
        }else{
            newShowView.index = self.showViewArr.count;
            [self.showViewArr addObject:newShowView];
        }
        // 加入字典
        [self.showViewDict setObject:newShowView forKey:[self getDictKey:showModel]];
        
    }
    // 如果存在旧模型
    else{
        // 修改数量大小
        if (isResetNumber) {
            [oldShowView resetTimeAndNumberFrom:showNumber];
        }else{
            [oldShowView addGiftNumberFrom:1];
        }
        if (self.railwayCanExchange) {
            // 比较数量大小排序
            [self sortShowArr];
            // 排序后调整Y值
            [self resetY];
        }
    }
}

- (void)animatedWithGiftModel:(LiveGiftShowModel *)showModel {
    if (!showModel) { return ;}
    // 线程安全：统一转到主队列
    if (![NSThread isMainThread]) {
        __weak __typeof(self)weakSelf = self;
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf animatedWithGiftModel:showModel];
        });
        return;
    }

    // 同一弹幕已有连击定时器在跑时，合并次数而不是再开一个定时器。
    // 多个并发定时器同时驱动同一个视图自增，是数字异常膨胀（issue #17）的根源。
    LiveGiftShowView * showingView = [self.showViewDict objectForKey:[self getDictKey:showModel]];
    if (showingView.model.animatedTimer != nil && showingView.model != showModel) {
        showingView.model.toNumber += showModel.toNumber;
        return;
    }

    if (self.addMode == LiveGiftAddModeQueue) {
        // 不存在旧视图
        if (!showingView) {
            NSUInteger showCount = 0;
            for (id object in self.showViewArr) {
                if ([object isKindOfClass:[LiveGiftShowView class]]) {
                    showCount ++;
                }
            }
            // 弹幕数量大于最大数量
            if (showCount >= self.maxRailwayCount) {
                [self addToQueue:showModel];
                return;
            }
        }
    }

    // 当前定时器源
    dispatch_source_t tt = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, dispatch_get_main_queue());
    // 任务执行开始时间
    dispatch_time_t start = dispatch_walltime(NULL, 0);
    // 任务执行间隔时间
    uint64_t interval = showModel.interval * NSEC_PER_SEC;
    // 给定时器源绑定开始时间、间隔时间以及容忍误差时间
    // 容差取间隔的 10%，允许系统合并定时器唤醒，降低多弹幕连击时的瞬时 CPU
    dispatch_source_set_timer(tt, start, interval, interval / 10);

    // 创建即持有，保证合并判断（animatedTimer != nil）在首个 tick 之前就生效
    showModel.animatedTimer = tt;

    __block NSInteger i = 0;
    __weak __typeof(self)weakSelf = self;
    // 给定时器源绑定任务
    dispatch_source_set_event_handler(tt, ^{
        if (!weakSelf.superview) {
            dispatch_cancel(tt);
            showModel.animatedTimer = nil;
            return;
        }
        if (i < showModel.toNumber) {
            i ++;
            [weakSelf addLiveGiftShowModel:showModel];
        } else {
            dispatch_cancel(tt);
            // 置 nil 让后续同 key 的连击能重新开定时器 / 队列合并逻辑能正确判断
            showModel.animatedTimer = nil;
        }
    });

    // 启动定时器源
    dispatch_resume(tt);
}

- (void)appearWith:(LiveGiftShowView *)showView {
    // 出现的动画
    if (self.appearMode == LiveGiftAppearModeLeft) {
        showView.isAppearAnimation = YES;
        [UIView animateWithDuration:self.kAppearAnimationTime animations:^{
            CGRect f = showView.frame;
            f.origin.x = 0;
            showView.frame = f;
        } completion:^(BOOL finished) {
            showView.isAppearAnimation = NO;
        }];
    }
}

- (void)resetY {
//    NSLog(@"重置 Y 轴了");
    for (int i = 0; i < self.showViewArr.count; i++) {
        LiveGiftShowView * show = self.showViewArr[i];
        if ([show isKindOfClass:[LiveGiftShowView class]]) {
            CGFloat showY = i * (kViewHeight+kGiftViewMargin);
            if (self.showMode == LiveGiftShowModeFromBottomToTop) {
                showY = -showY;
            }
            if (show.frame.origin.y != showY) {
                if (!show.isLeavingAnimation) {
                    // 避免出现动画和交换动画冲突
                    if (show.isAppearAnimation) {
                        [show.layer removeAllAnimations];
                    }
                    [UIView animateWithDuration:self.kExchangeAnimationTime animations:^{
                        CGRect showF = show.frame;
                        showF.origin.y = showY;
                        show.frame = showF;
                    } completion:^(BOOL finished) {
                        
                    }];
                    show.isAnimation = YES;
                    if ([self isDebug]) {
                        WLog(@"%@ 重置动画",show);
                    }
                }
            }
        }
    }
}

- (void)sortShowArr{
    // 如果当前数组包含kGiftViewRemoved 则将kGiftViewRemoved替换到LiveGiftShowView之后
    for (int i = 0; i < self.showViewArr.count; i++) {
        id current = self.showViewArr[i];
        if ([current isKindOfClass:[NSString class]]){
            if (i+1 < self.showViewArr.count) {
                [self searchLiveShowViewFrom:i+1];
            }
        }
    }
    // 以当前 数字大小 比较，降序
    for (int i = 0; i < self.showViewArr.count; i++) {
        for (int j = i; j < self.showViewArr.count; j++) {
            LiveGiftShowView * showViewI = self.showViewArr[i];
            LiveGiftShowView * showViewJ = self.showViewArr[j];
            if ([showViewI isKindOfClass:[LiveGiftShowView class]] && [showViewJ isKindOfClass:[LiveGiftShowView class]]) {
                if ([showViewI.numberView currentNumber] < [showViewJ.numberView currentNumber]) {
                    showViewI.index = j;
                    showViewI.isAnimation = YES;
                    showViewJ.index = i;
                    showViewJ.isAnimation = YES;
                    [self.showViewArr exchangeObjectAtIndex:i withObjectAtIndex:j];
                }
            }
        }
    }
    if ([self isDebug]) {
//        WLog(@"排序后数组==>>> %@",self.showViewArr);
    }
}

- (void)searchLiveShowViewFrom:(int)i{
    for (int j = i; j < self.showViewArr.count; j++) {
        LiveGiftShowView * next = self.showViewArr[j];
        if ([next isKindOfClass:[LiveGiftShowView class]]) {
            // 正在飞出的视图不参与补位（原实现用 frame.origin.x == 屏幕宽度 的浮点相等判断，
            // 只覆盖右出场景且依赖动画结束时机，改用语义明确的状态位）
            if (next.isLeavingAnimation) {
                continue;
            }
            next.index = i-1;
            [self.showViewArr exchangeObjectAtIndex:i-1 withObjectAtIndex:j];
            return;
        }
    }
}

#pragma mark - Deprecated（V2.0 起请直接使用同名属性）

- (void)enableInterfaceDebug:(BOOL)isDebug {
    self.interfaceDebugEnabled = isDebug;
}

- (void)setHiddenModel:(LiveGiftHiddenMode)model {
    self.hiddenMode = model;
}

- (void)setAppearModel:(LiveGiftAppearMode)model {
    self.appearMode = model;
}

- (BOOL)isDebug {
    return self.interfaceDebugEnabled;
}

#pragma mark - Private
- (void)addToQueue:(LiveGiftShowModel *)showModel {
    if (!showModel) {
        return;
    }
    NSString * key = [self getDictKey:showModel];
    NSUInteger oldNumber = 0;
    for (NSUInteger i = 0; i<self.waitQueueArr.count; i++) {
        LiveGiftShowModel * oldModel = self.waitQueueArr[i];
        NSString * oldKey = [self getDictKey:oldModel];
        if ([oldKey isEqualToString:key] && oldModel.animatedTimer == nil) {
            oldNumber = oldModel.toNumber;
            showModel.toNumber += oldNumber;
            [self.waitQueueArr removeObject:oldModel];
            break;
        }
    }
    [self.waitQueueArr addObject:showModel];
}

/// 展示等待队列
- (void)showWaitView {
    NSUInteger showCount = 0;
    for (id object in self.showViewArr) {
        if ([object isKindOfClass:[LiveGiftShowView class]]) {
            showCount ++;
        }
    }
    if (showCount < self.maxRailwayCount) {
        LiveGiftShowModel * model = self.waitQueueArr.firstObject;
        if (model.currentNumber > 0) {
            [self addLiveGiftShowModel:model showNumber:model.currentNumber];
        } else {
            [self animatedWithGiftModel:model];
        }
        [self.waitQueueArr removeObject:model];
    }
}

- (void)resetView:(LiveGiftShowView *)view nowModel:(LiveGiftShowModel *)model isChangeNum:(BOOL)isChange number:(NSInteger)number{
    NSString * oldKey = [self getDictKey:view.model];
    NSString * dictKey = [self getDictKey:model];
    
    if (view.model.animatedTimer) {
        dispatch_cancel(view.model.animatedTimer);
        view.model.animatedTimer = nil;
    }
    //找到时间早的那个视图 替换模型 重置数字
    view.imageLoader = self.webImageLoader;
    view.model = model;
    if (isChange) {
        [view resetTimeAndNumberFrom:number];
    }else{
        [view resetTimeAndNumberFrom:1];
    }
    [self.showViewDict removeObjectForKey:oldKey];
    [self.showViewDict setObject:view forKey:dictKey];
}

- (NSString *)getDictKey:(LiveGiftShowModel *)model{
    // 优先以 userId+礼物类型 为 key，避免同名用户的弹幕被错误合并；未传 userId 时退化为用户名
    NSString * userKey = model.user.userId.length > 0 ? model.user.userId : model.user.name;
    NSString * key = [NSString stringWithFormat:@"%@%@",userKey,model.giftModel.type];
    return key;
}

#pragma mark - Lazy
- (NSMutableArray<LiveGiftShowModel *> *)waitQueueArr {
    if (!_waitQueueArr) {
        _waitQueueArr = [[NSMutableArray alloc]init];
    }
    return _waitQueueArr;
}

- (NSMutableDictionary *)showViewDict{
    if (!_showViewDict) {
        _showViewDict = [[NSMutableDictionary alloc]init];
    }
    return _showViewDict;
}

-(NSMutableArray *)showViewArr{
    if (!_showViewArr) {
        _showViewArr = [[NSMutableArray alloc]init];
    }
    return _showViewArr;
}

- (void)dealloc{
    if ([self isDebug]) {
        WLog(@"Delloc LiveGiftShowCustom !!  %@",self);
    }
}


@end

File: Sources/LiveSendGiftSwift/LiveGiftShowCustom.swift
//
//  LiveGiftShowCustom.swift
//  LiveSendGiftSwift
//
//  管理所有弹幕的容器视图。
//  行为与 ObjC 版 LiveGiftShowCustom（V2.0）保持一致：
//  实例级配置、线程安全、userId 区分用户、同 key 连击定时器合并。
//

import UIKit

@objc(SwiftLiveGiftShowCustom)
public final class LiveGiftShowCustom: UIView {

    /// 两个弹幕之间的高度差
    static let giftViewMargin: CGFloat = 50.0

    // MARK: - 配置（实例属性，多实例互不影响）

    /// 交换动画时长
    public var exchangeAnimationTime: TimeInterval = 0.25
    /// 出现时动画时长
    public var appearAnimationTime: TimeInterval = 0.5
    /// 弹幕添加模式，默认替换
    public var addMode: LiveGiftAddMode = .replace
    /// 最大礼物轨道数量，默认 3
    public var maxRailwayCount: Int = 3 {
        didSet {
            heightConstraint?.constant =
                (kLiveGiftViewHeight + Self.giftViewMargin) * CGFloat(maxRailwayCount - 1) + kLiveGiftViewHeight
        }
    }
    /// 轨道能否进行交换动画，默认 true
    public var railwayCanExchange = true
    /// 弹幕展现模式，默认自上而下
    public var showMode: LiveGiftShowMode = .fromTopToBottom
    /// 弹幕消失模式，默认向右移出
    public var hiddenMode: LiveGiftHiddenMode = .right
    /// 弹幕出现模式，默认从左出现
    public var appearMode: LiveGiftAppearMode = .left
    /// 是否打印调试信息，默认 false
    public var interfaceDebugEnabled = false

    /// 网络图片加载器。默认 nil：使用内置的 URLSession 加载器。
    /// 使用 Kingfisher/SDWebImage/自研加载器的宿主可注入此 block。
    public var webImageLoader: LiveGiftWebImageLoader?

    /// 弹幕被移除时的回调
    public var onGiftRemoved: ((LiveGiftShowModel) -> Void)?

    // MARK: - 内部状态

    /// 轨道槽位，nil 表示该轨道已空（原 ObjC 版的 kGiftViewRemoved 占位）
    private(set) var slots: [LiveGiftShowView?] = []
    /// key（userId/name + 礼物类型) -> 正在展示的弹幕
    private(set) var showViewDict: [String: LiveGiftShowView] = [:]
    /// 待展示礼物队列
    private(set) var waitQueue: [LiveGiftShowModel] = []

    private var heightConstraint: NSLayoutConstraint?

    // MARK: - 初始化

    /// 添加到宿主视图。y 的设定应注意最大礼物数量时不要超出屏幕边界，
    /// 建议按安全区计算（如 `view.safeAreaInsets.top + 10`）。
    @discardableResult
    public static func add(to superView: UIView, y: CGFloat = 100) -> LiveGiftShowCustom {
        let v = LiveGiftShowCustom()
        v.isUserInteractionEnabled = false
        v.backgroundColor = .clear
        superView.addSubview(v)
        v.translatesAutoresizingMaskIntoConstraints = false
        let height = v.heightAnchor.constraint(
            equalToConstant: (kLiveGiftViewHeight + giftViewMargin) * CGFloat(v.maxRailwayCount - 1)
                + kLiveGiftViewHeight)
        v.heightConstraint = height
        NSLayoutConstraint.activate([
            v.widthAnchor.constraint(equalToConstant: kLiveGiftViewWidth),
            height,
            v.leftAnchor.constraint(equalTo: superView.leftAnchor),
            v.topAnchor.constraint(equalTo: superView.topAnchor, constant: y),
        ])
        return v
    }

    // MARK: - 公开 API（线程安全：非主线程调用自动转主队列）

    /// 增加或者更新一个礼物弹幕。
    /// - Parameters:
    ///   - showModel: 礼物模型
    ///   - showNumber: 大于 0 时直接显示该数字，否则从 1 开始自增
    public func add(_ showModel: LiveGiftShowModel, showNumber: Int = 0) {
        guard Thread.isMainThread else {
            DispatchQueue.main.async { [weak self] in self?.add(showModel, showNumber: showNumber) }
            return
        }

        let key = dictKey(showModel)
        let isResetNumber = showNumber > 0

        // 已有同 key 弹幕：更新计数
        if let oldView = showViewDict[key] {
            if isResetNumber {
                oldView.resetTimeAndNumber(from: showNumber)
            } else {
                oldView.addGiftNumber(from: 1)
            }
            if railwayCanExchange {
                compactAndSort()
                resetY()
            }
            return
        }

        // 轨道已满且无空位
        if slots.count >= maxRailwayCount, !slots.contains(where: { $0 == nil }) {
            switch addMode {
            case .replace:
                // 替换创建时间最早的弹幕
                if let oldest = slots.compactMap({ $0 }).min(by: { $0.createDate < $1.createDate }) {
                    reset(oldest, with: showModel, isChangeNumber: isResetNumber, number: showNumber)
                }
            case .queue:
                if showNumber > 0 {
                    showModel.currentNumber = showNumber
                }
                addToQueue(showModel)
            }
            return
        }

        // 计算轨道与 Y 值
        let slotIndex = slots.firstIndex(where: { $0 == nil }) ?? slots.count
        var showViewY = CGFloat(slotIndex) * (kLiveGiftViewHeight + Self.giftViewMargin)
        if showMode == .fromBottomToTop {
            showViewY = -showViewY
        }
        var frame = CGRect(x: 0, y: showViewY, width: 0, height: 0)
        if appearMode == .left {
            // 用宿主视图宽度计算离屏起点，iPad 分屏等非全屏容器下 mainScreen 宽度会偏大
            let containerWidth = superview?.bounds.width ?? UIScreen.main.bounds.width
            frame.origin.x = -containerWidth
        }

        let newView = LiveGiftShowView(frame: frame)
        // imageLoader 须在 model 之前，didSet 内会触发图片加载
        newView.imageLoader = webImageLoader
        newView.model = showModel
        newView.hiddenMode = hiddenMode
        if isResetNumber {
            newView.resetTimeAndNumber(from: showNumber)
        } else {
            newView.addGiftNumber(from: 1)
        }
        appear(newView)

        // 超时移除
        newView.onTimeOut = { [weak self] view in
            guard let self else { return }
            if let model = view.model {
                self.onGiftRemoved?(model)
            }
            // 释放槽位（index 由排序逻辑维护，做防护避免状态不同步）
            if view.index >= 0, view.index < self.slots.count, self.slots[view.index] === view {
                self.slots[view.index] = nil
            } else if let realIndex = self.slots.firstIndex(where: { $0 === view }) {
                self.slots[realIndex] = nil
            }
            if let model = view.model {
                self.showViewDict.removeValue(forKey: self.dictKey(model))
            }
            self.debugLog("移除了第 \(view.index) 个，剩余 \(self.showViewDict.count) 条")

            if self.railwayCanExchange {
                self.compactAndSort()
                self.resetY()
            }
            switch self.addMode {
            case .queue:
                self.showWaitView()
            case .replace:
                if let timer = view.model?.animatedTimer {
                    timer.cancel()
                    view.model?.animatedTimer = nil
                }
            }
        }

        addSubview(newView)

        // 占用槽位
        if slotIndex < slots.count {
            newView.index = slotIndex
            slots[slotIndex] = newView
        } else {
            newView.index = slots.count
            slots.append(newView)
        }
        showViewDict[key] = newView
    }

    /// 添加一个礼物弹幕：若该礼物不在视图上，则播放从 1 增加到 toNumber 的连击效果；否则继续累加。
    public func animate(with showModel: LiveGiftShowModel) {
        guard Thread.isMainThread else {
            DispatchQueue.main.async { [weak self] in self?.animate(with: showModel) }
            return
        }

        // 同一弹幕已有连击定时器在跑时，合并次数而不是再开一个定时器。
        // 多个并发定时器同时驱动同一个视图自增，是数字异常膨胀（ObjC 版 issue #17）的根源。
        let key = dictKey(showModel)
        if let showing = showViewDict[key], let runningModel = showing.model,
            runningModel.animatedTimer != nil, runningModel !== showModel
        {
            runningModel.toNumber += showModel.toNumber
            return
        }

        if addMode == .queue, showViewDict[key] == nil {
            let showCount = slots.compactMap { $0 }.count
            if showCount >= maxRailwayCount {
                addToQueue(showModel)
                return
            }
        }

        let timer = DispatchSource.makeTimerSource(queue: .main)
        // 容差取间隔的 10%，允许系统合并定时器唤醒，降低多弹幕连击时的瞬时 CPU
        timer.schedule(
            deadline: .now(),
            repeating: showModel.interval,
            leeway: .milliseconds(Int(showModel.interval * 100)))
        // 创建即持有，保证合并判断（animatedTimer != nil）在首个 tick 之前就生效
        showModel.animatedTimer = timer

        var tick = 0
        timer.setEventHandler { [weak self] in
            guard let self, self.superview != nil else {
                timer.cancel()
                showModel.animatedTimer = nil
                return
            }
            if tick < showModel.toNumber {
                tick += 1
                self.add(showModel)
            } else {
                timer.cancel()
                // 置 nil 让后续同 key 的连击能重新开定时器 / 队列合并逻辑能正确判断
                showModel.animatedTimer = nil
            }
        }
        timer.resume()
    }

    // MARK: - Private

    /// 出现动画
    private func appear(_ view: LiveGiftShowView) {
        guard appearMode == .left else { return }
        view.isAppearAnimation = true
        UIView.animate(
            withDuration: appearAnimationTime,
            animations: {
                var f = view.frame
                f.origin.x = 0
                view.frame = f
            },
            completion: { _ in
                view.isAppearAnimation = false
            })
    }

    /// 空槽位补位 + 按当前计数降序排序
    private func compactAndSort() {
        // 空槽位由后续未在飞出的弹幕补位
        for p in 0..<slots.count where slots[p] == nil {
            for j in (p + 1)..<slots.count {
                if let view = slots[j], !view.isLeavingAnimation {
                    view.index = p
                    slots.swapAt(p, j)
                    break
                }
            }
        }
        // 计数大的弹幕排在上面
        for i in 0..<slots.count {
            for j in i..<slots.count {
                guard let viewI = slots[i], let viewJ = slots[j],
                    viewI.numberView.currentNumber < viewJ.numberView.currentNumber
                else { continue }
                viewI.index = j
                viewI.isAnimating = true
                viewJ.index = i
                viewJ.isAnimating = true
                slots.swapAt(i, j)
            }
        }
    }

    /// 排序后调整各弹幕的 Y 值
    private func resetY() {
        for (i, slot) in slots.enumerated() {
            guard let show = slot else { continue }
            var showY = CGFloat(i) * (kLiveGiftViewHeight + Self.giftViewMargin)
            if showMode == .fromBottomToTop {
                showY = -showY
            }
            guard show.frame.origin.y != showY, !show.isLeavingAnimation else { continue }
            // 避免出现动画和交换动画冲突
            if show.isAppearAnimation {
                show.layer.removeAllAnimations()
            }
            UIView.animate(withDuration: exchangeAnimationTime) {
                var f = show.frame
                f.origin.y = showY
                show.frame = f
            }
            show.isAnimating = true
            debugLog("\(show) 重置动画")
        }
    }

    /// 替换模式：复用最早的弹幕视图展示新模型
    private func reset(_ view: LiveGiftShowView, with model: LiveGiftShowModel, isChangeNumber: Bool, number: Int) {
        if let oldModel = view.model {
            showViewDict.removeValue(forKey: dictKey(oldModel))
            if let timer = oldModel.animatedTimer {
                timer.cancel()
                oldModel.animatedTimer = nil
            }
        }
        view.imageLoader = webImageLoader
        view.model = model
        view.resetTimeAndNumber(from: isChangeNumber ? number : 1)
        showViewDict[dictKey(model)] = view
    }

    /// 加入等待队列；同 key 且未在连击中的等待模型会被合并
    private func addToQueue(_ model: LiveGiftShowModel) {
        let key = dictKey(model)
        if let index = waitQueue.firstIndex(where: { dictKey($0) == key && $0.animatedTimer == nil }) {
            model.toNumber += waitQueue[index].toNumber
            waitQueue.remove(at: index)
        }
        waitQueue.append(model)
    }

    /// 展示等待队列中的下一条
    private func showWaitView() {
        let showCount = slots.compactMap { $0 }.count
        guard showCount < maxRailwayCount, let model = waitQueue.first else { return }
        waitQueue.removeFirst()
        if model.currentNumber > 0 {
            add(model, showNumber: model.currentNumber)
        } else {
            animate(with: model)
        }
    }

    /// 弹幕归属 key：优先 userId，避免同名用户被错误合并
    func dictKey(_ model: LiveGiftShowModel) -> String {
        let userKey: String
        if let userId = model.user.userId, !userId.isEmpty {
            userKey = userId
        } else {
            userKey = model.user.name ?? ""
        }
        return userKey + (model.gift.type ?? "")
    }

    private func debugLog(_ message: @autoclosure () -> String) {
        #if DEBUG
        if interfaceDebugEnabled {
            print("[LiveSendGiftSwift] \(message())")
        }
        #endif
    }
}
